src/BlockList.h

```diff
@@ -0,0 +1,479 @@
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+/**
+ * A concurrent list data structure, which is implemented through a chunked linked-list.
+ * On their own, get/push_back are threadsafe, however clearing/destructing is undefined
+ * behaviour if get/push_backs are in progress.
+ */
+template <class T>
+class concurrent_list {
+    // how large each cv_data->m_data array is
+    const size_t chunk_size = 1000;
+
+    struct cv_data {
+        // how many elements are currently in m_data
+        std::atomic<size_t> m_size;
+        T* m_data;
+        // ptr to the next cv_data
+        std::atomic<cv_data*> next;
+    };
+
+    // number of elements in total
+    std::atomic<size_t> container_size;
+
+    // only necessary for those that change vectors
+    mutable std::mutex writeMutex;
+
+    std::atomic<cv_data*> cData;
+
+public:
+    concurrent_list() {
+        cv_data* curr = new cv_data;
+        curr->m_size = 0;
+        curr->m_data = new T[chunk_size];
+        // let's just fill with empty for test
+        // for (int  i = 0; i < chunk_size; ++i) curr->m_data[i] = nullptr;
+        curr->next = nullptr;
+        cData = curr;
+        container_size = 0;
+    }
+
+    concurrent_list(concurrent_list&& other) : concurrent_list() {
+        std::unique_lock<std::mutex> lk1(other.writeMutex);
+        std::unique_lock<std::mutex> lk2(writeMutex);
+
+        size_t tmp = this->container_size.load();
+        this->container_size.store(other.container_size);
+        other.container_size.store(tmp);
+
+        cv_data* tmpPtr = this->cData.load();
+        this->cData.store(other.cData);
+        other.cData.store(tmpPtr);
+
+        // std::swap(old.container_size, container_size);
+        // std::swap(old.cData, cData);
+    }
+
+    concurrent_list& operator=(concurrent_list other) {
+        size_t tmp = this->container_size.load();
+        this->container_size.store(other.container_size);
+        other.container_size.store(tmp);
+
+        cv_data* tmpPtr = this->cData.load();
+        this->cData.store(other.cData);
+        other.cData.store(tmpPtr);
+
+        return *this;
+    }
+
+    /* it goes without saying this is not threadsafe */
+    ~concurrent_list() {
+        // lock just to let writes finish
+        std::unique_lock<std::mutex> lk(writeMutex);
+
+        auto curr = cData.load();
+        while (curr != nullptr) {
+            cv_data* tmpnext = curr->next;
+            delete[] curr->m_data;
+            delete curr;
+
+            curr = tmpnext;
+        }
+    }
+
+    /**
+     * the concept of size() in a threaded environment is peculiar
+     * However, it is guaranteed that the size returned is a valid index+1
+     * if the data structure has not been concurrently reduced.
+     */
+    size_t size() const {
+        return container_size;
+    }
+
+    /* thread safe for many writers, although in practise, it is sequential through mutex */
```

Comment:
I left that as a comment for me to come back and possibly partition the function to not be totally blocking. I've replaced the doc string with something less ambiguous.